spring-cloud-gcp-starters/spring-cloud-gcp-starter-trace/src/main/java/org/springframework/cloud/gcp/trace/autoconfig/StackdriverTraceAutoConfiguration.java

```diff
@@ -0,0 +1,166 @@
+import com.google.cloud.trace.v1.util.RoughTraceSizer;
+import com.google.cloud.trace.v1.util.Sizer;
+import com.google.devtools.cloudtrace.v1.Trace;
+
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.boot.autoconfigure.AutoConfigureBefore;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cloud.commons.util.IdUtils;
+import org.springframework.cloud.gcp.core.GcpProjectIdProvider;
+import org.springframework.cloud.gcp.trace.GcpTraceProperties;
+import org.springframework.cloud.gcp.trace.TraceServiceClientTraceConsumer;
+import org.springframework.cloud.gcp.trace.sleuth.LabelExtractor;
+import org.springframework.cloud.gcp.trace.sleuth.StackdriverTraceSpanListener;
+import org.springframework.cloud.sleuth.Sampler;
+import org.springframework.cloud.sleuth.SpanAdjuster;
+import org.springframework.cloud.sleuth.SpanNamer;
+import org.springframework.cloud.sleuth.SpanReporter;
+import org.springframework.cloud.sleuth.TraceKeys;
+import org.springframework.cloud.sleuth.autoconfig.TraceAutoConfiguration;
+import org.springframework.cloud.sleuth.log.SpanLogger;
+import org.springframework.cloud.sleuth.sampler.PercentageBasedSampler;
+import org.springframework.cloud.sleuth.sampler.SamplerProperties;
+import org.springframework.cloud.sleuth.trace.DefaultTracer;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Import;
+import org.springframework.context.annotation.Primary;
+import org.springframework.core.env.Environment;
+
+@Configuration
+@EnableConfigurationProperties({ SamplerProperties.class })
+@ConditionalOnProperty(value = "spring.cloud.gcp.trace.enabled", matchIfMissing = true)
+@Import({ StackdriverTraceAutoConfiguration.TraceConsumerConfiguration.class })
+@AutoConfigureBefore(TraceAutoConfiguration.class)
+public class StackdriverTraceAutoConfiguration {
+	@Bean
+	@ConditionalOnMissingBean
+	public Sampler defaultTraceSampler(SamplerProperties config) {
+		return new PercentageBasedSampler(config);
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	public Random randomForSpanIds() {
+		return new SecureRandom();
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	public SpanReporter traceSpanReporter(Environment environment, GcpProjectIdProvider gcpProjectIdProvider,
+			LabelExtractor labelExtractor,
+			List<SpanAdjuster> spanAdjusters, TraceConsumer traceConsumer) {
+		String instanceId = IdUtils.getDefaultInstanceId(environment);
+		return new StackdriverTraceSpanListener(instanceId, gcpProjectIdProvider.getProjectId(), labelExtractor,
+				spanAdjusters, traceConsumer);
+	}
+
+	@Bean
+	@ConditionalOnMissingBean(Tracer.class)
```

Comment:
Could this method just return a `Tracer`? It's not immediately clear to me why it can't.

Reply:
I was wondering that too. But decided to follow the convention here: https://github.com/spring-cloud/spring-cloud-sleuth/blob/master/spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/autoconfig/TraceAutoConfiguration.java

Reply:
I have no idea why we put `DefaultTracer` there ;) It should be `Tracer`.